### ISSUE TITLE
Add support for evaluator functions

### ISSUE THREAD

**Issue by tohrnii** [0xPolygonMiden/air-script#124]
We can introduce "evaluator" functions which would take the description of main and auxiliary traces as parameters which would allow us to do something like:

```
ev foo(main: [a, b, ..2], aux: [c, d]):
  enf a' = a + b
```

One advantage of this approach is that we can name evaluator functions and then invoke them from other places. For example:

```
ev foo(main: [a, b, ..2], aux: [c, d]):
  enf bar(main[0..2])
  // some other constraints

ev bar(main: [a, b]):
  enf a' = a + b
```

Rescue Prime constraints in this format could look something like this:

```
rescue.air
--------------------
cycl ark1 = [
  /* 12 periodic column definitions */
]
cycl ark2 = [
  /* 12 periodic column definitions */
]

const mds = [
  /* 12x12 matrix */
]
const inv_mds = [
  /* 12x12 matrix */
]

#[prover("mds.rs")]
fn apply_mds(state: vector[12]) -> vector[12]:
  [
    sum([m * s for (m, s) in (mds_row, state)])
    for mds_row in mds
  ]

fn apply_inv_mds(state: vector[12]) -> vector[12]:
  [
    sum([m * s for (m, s) in (inv_mds_row, state)])
    for inv_mds_row in inv_mds
  ]

ev rescue_prime(main: [state[12]]):
  let s1 = [x^7 for x in state]
  let s1 = apply_mds(s1)
  let s1 = [x + k for (x, k) in (s1, ark1)]

  let s2 = [x' + k for (x, k) in (state, ark2)]
  let s2 = apply_inv_mds(s2)
  let s2 = [x^7 for x in s2]

  enf x = y for (x, y) in (s1, s2)
```

And then we could invoke them from another constraint file like so:

```
hasher.air
--------------------
use rescue::rescue_prime

ev hasher(main: [s[3], r, h[12], idx]):
  enf rescue_prime(h)
```

_Originally posted by @bobbinth in https://github.com/0xPolygonMiden/miden-vm/discussions/254#discussioncomment-3302565_

**Comment by bobbinth** [comment]
It may make sense to separate this into two issues:
1. Evaluator functions in a single file.
2. Invoking evaluator functions from imported files.

**Comment by grjte** [comment]
closed by #227  and #260
